packages/flutter/lib/src/widgets/sliver.dart

```diff
@@ -966,35 +1006,89 @@ class SliverMultiBoxAdaptorElement extends RenderObjectElement implements Render
 
   final Map<int, Widget> _childWidgets = HashMap<int, Widget>();
   final SplayTreeMap<int, Element> _childElements = SplayTreeMap<int, Element>();
-  RenderBox _currentBeforeChild;
+
 
   @override
   void performRebuild() {
+    renderObject.integrityEnabled = false; // moving children will temporary violate the integrity
     _childWidgets.clear(); // Reset the cache, as described above.
     super.performRebuild();
-    _currentBeforeChild = null;
     assert(_currentlyUpdatingChildIndex == null);
     try {
+      final Set<int> affectedIndexes = {};
+      final Map<int, Element> newKeepAliveChildren = <int, Element>{};
+
+      int getNewIndex(Widget w){
+        if (w.key != null){
+          return widget.delegate.getIndexByKey(w.key);
+        } else {
+          return -1;
+        }
+      }
+
       void processElement(int index) {
         _currentlyUpdatingChildIndex = index;
         final Element newChild = updateChild(_childElements[index], _build(index), index);
         if (newChild != null) {
           _childElements[index] = newChild;
-          final SliverMultiBoxAdaptorParentData parentData = newChild.renderObject.parentData;
-          if (!parentData.keptAlive)
-            _currentBeforeChild = newChild.renderObject;
         } else {
           _childElements.remove(index);
         }
       }
-      // processElement may modify the Map - need to do a .toList() here.
-      _childElements.keys.toList().forEach(processElement);
+
```

Comment:
I wonder if the following logic could be simplified a bit (with less list allocations and list iterations) if we first build a map that maps all indexes we want to fill out to the old Element that's supposed to get reused for that index (or `null` if there is no Element to re-use for that index). Once we have that map, we could take the keys (= indexes), sort them, and call a slightly modified `processElement` on them that calls updateChild with the old element from the map.

Reply:
We could even do that "in order", which may mean that we can re-use the existing logic to determine the "insert-after" child...

Reply:
I think I can combine them and and do them in order. but what do you think of the _findAfterBySlot method? I feel it might be more secure in some cases, although it is built in the assumption on how keptalivebucket work.